feat(test): protractor integration, e2e test sample

filipesilva opened the pull request:
Fix https://github.com/angular/angular-cli/issues/72

cironunes commented on README.md:
```diff
@@ -64,19 +64,21 @@ The build artifacts will be stored in the `dist/` directory.
+This will be easier when the command
 [ng test](https://github.com/angular/angular-cli/issues/70) is implemented.
 
+To run the end-to-end protractor tests, just do `$(npm bin)/protractor`.
```
Please add a note saying that you need to have the server running (`ng serve`) to execute protractor tests?

filipesilva replied:
Roger, it is done.

cironunes commented on addon/ng2/blueprints/ng2/files/package.json:
```diff
@@ -2,6 +2,9 @@
   "version": "0.0.0",
   "license": "Apache-2.0",
+  "scripts":{
+    "postinstall": "webdriver-manager update"
```
This is good to have, but it's slowing down a lot the init time. It would be good to start measuring it and trying to cut it down.

/cc @IgorMinar

filipesilva replied:
We can run it right before protractor instead, as part of the readme instructions. Afterwards, as part of `ng test` if protractor is integrated with it.

cironunes replied:
Yeah. I think that we should make an instruction on readme instead. We don't need to update webdriver for each project that we create...

filipesilva replied:
I actually do have a question about that. I'm partial to all-local dependencies (bower/karma/protractor) because that way each project can be version locked, and would generally only need node installed to run (and no global packages at all). Doing things this way will undoubtedly add time to the initial startup though.

So right now, a project generated in this way will have it's own protractor and own webdriver, thus the `webdriver-manager update` is necessary to be ran for each project.

What's your view on this?

cironunes replied:
You're right. That's actually the way we're doing.

I thought for a moment that the webdriver version was independent of Protractor, but that doesn't make sense.

We still need to try to make our init time faster. Running `webdriver-manager update` as part of `ng test` would be nice for the first run, but it would be unnecessary for the subsequent runs. What else we can do?

filipesilva replied:
Hm..... perhaps we can do something more involved in `ng test`. Like only running `webdriver-manager update` if there is no driver at all. 

I checked and `webdriver-manager` api doesn't seem to have something to do that. Perhaps I could do a PR to implement such functionality in protractor itself, a sort of `webdriver-manager initialize`.

@juliemr do you think this would be reasonable?

juliemr replied:
I'm not entirely sure what you're proposing. `webdriver-manager update` will only actually download stuff if it finds it needs new files, so it already does this? Unless you mean something different by "no driver at all" - like looking in some shared location?

I know the slow download time is a pain point, so I'm happy to take ideas and figure out a better solution overall.

filipesilva replied:
I was also under the erroneous impression that `webdriver-manager update` would re-download! Just tested it and it just ain't so. My bad, sorry for wasting your time @juliemr :/

@cironunes this means that we can just run it within `ng test` and it should be perfectly fine.

cironunes commented:
Thanks for the PR @filipesilva. I added a few comments, can you check?

filipesilva commented:
Let me know what you'd like to do about `webdriver` and I'll add that in.

In the meanwhile, do you have something else specific for me to do? Or should I just go over `help-wanted` issues/

cironunes commented:
@filipesilva you can take any issue that you feel like taking. The focus now is to build a cross-platform e2e testing suite for the project, but any help would be appreciated.

filipesilva commented:
I moved the `webdriver-manager update` command to the readme. Anything else you think should be changed?

cironunes commented:
@filipesilva let's just figure out how we'll integrate this with `ng test` so we can get this merged.

filipesilva commented:
Hm... although `ng test` should be the umbrella command, unit tests and e2e have different use cases. 

One would either run the unit tests once just to check everything is fine, or leave them running continuously while coding. In contrast, e2e tests are (in my experience) only ran once to check everything is fine. 

Thus it might make sense to run e2e tests while running unit tests in single run, but it does not make sense to run e2e when unit tests are not in single run.

So to start off, I'd suggest that e2e tests are only ran explicitly, via `ng test --e2e` or something similar.

IgorMinar commented:
@filipesilva right on - see my comment: https://github.com/angular/angular-cli/pull/86#issuecomment-163516453

IgorMinar commented on addon/ng2/blueprints/ng2/files/src/app/__name__.e2e.ts:
```diff
@@ -0,0 +1,9 @@
+describe('<%= jsComponentName %> App', function() {
+
+  // #docregion redirect
```
docregion? :)

filipesilva replied:
Ah, a remnant of the devguide example, my bad :p

IgorMinar commented on addon/ng2/blueprints/ng2/files/src/app/__name__.e2e.ts:
```diff
@@ -0,0 +1,9 @@
+  // #docregion redirect
+  it('should display message saying app works', function() {
+    browser.get('/');
+    expect(element(by.css('<%= htmlComponentName %>-app p')).getText())
```
don't we have a matcher in `angular2/testing` that makes this nicer?

filipesilva replied:
[toHaveText](https://github.com/angular/angular/blob/master/modules/angular2/src/testing/matchers.ts#L9) is the only one that seems applicable.

Trying to import anything from `import {} from 'angular2/testing';` seems to give me  `[launcher] Error: ReferenceError: System is not defined` in protractor. 

This seems to be because adding imports makes the generated files use `System`. Will have to look into how to make that work on protractor.

IgorMinar commented:
I'm not crazy about the manual steps we require, but we can fix that in follow up PRs.

How about renaming `e2e` extension to `e2eSpec`? Or is `e2e` clear enough? Maybe it is..

@juliemr do you have anything else you'd like to see in CLI when scaffolding the project/components?

filipesilva commented:
I'm partial to just `e2e`. I feel it's clear enough and doesn't lead users towards camelCase for their file names, eg `my-component.e2e.ts` vs `my-component.e2eSpec.ts`, and variations `my_component`, `myComponent`.

cironunes commented:
@filipesilva can you rebase the PR so we can get this merged?

filipesilva commented:
Apologies @cironunes, I had to return to docs for the last week.

This isn't ready to be merged yet though. While looking at the matchers that Igor was recommending, I found out since the `*.e2e.ts` files were being compiled to SystemJS module loading, that was causing all sorts of trouble when trying to use modules in general for protractor.

I will need to separate the e2e tests onto another folder, and provide it with `tsconfig.json` that uses CommonJS module loading.

filipesilva commented:
I've updated the commit to reflect the major changes, so please review again. 

Highlights:
- added `e2e` folder besides `src` with own `tsconfig.json`
- fixed imports
- added page object

cironunes commented:
@filipesilva can you rebase this with the latest changes from master?

filipesilva commented:
Rebased and confirmed it was still working.

cironunes commented:
lgtm
